lib/validators/restrict-children-validator.ts

```diff
@@ -0,0 +1,69 @@
+/**
+ * Validates restrictChildren properties option
+ */
+
+import { Validator } from './validator';
+import { DirectiveType, ParsedComponentsDefinition } from '../models';
+
+const PROPERTY = 'restrictChildren';
+const ADDITIONAL_PROPERTY = 'withContent';
+
+export class RestrictChildrenValidator implements Validator {
+
+    constructor(
+        private definition: ParsedComponentsDefinition,
+    ) {
+    }
+
+    private hasSlideshowDirective(parsedComponent: ParsedComponentsDefinition['components']['name']) : boolean {
+        return Object.values(parsedComponent.directives)
+        .some(directive => directive.type === DirectiveType.slideshow);
+    }
+
+    validate(
+        errorReporter: (errorMessage: string) => void,
+    ): boolean {
+        let valid = true;
+
+        Object.values(this.definition.components).forEach((parsedComponent: ParsedComponentsDefinition['components']['name']) => {
+            const isPresent = PROPERTY in parsedComponent.component && parsedComponent.component[PROPERTY];
+            if (!isPresent) {
+                if (this.hasSlideshowDirective(parsedComponent)) {
+                    errorReporter(`Component property "${PROPERTY}" must be defined in component "${parsedComponent.component.name}" because the ` + 
+                        `component contains slideshow directive`);
+                    valid = false;
+                }
+                return;
+            }
+            // check if all keys point to corrent component
+            // TODO if || {} is removed then ES2017 shows an error that parsedComponent.component[PROPERTY] may be undefined, why?
```

Comment:
Does anyone have an idea?

If strict mode is off (tsconfig) then it builds. Otherwise need to add that '|| {}'. With ES2015 it builds too.

Reply:
Guess Typescript can't infer that you already checked if the property exists?

Apparently there is an operator to assert that the value is non null (Non-null assertion operator):

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html

Reply:
I think Sander is right about TS not being able infer. Adding an `if` solves it (very ugly though). Upgrading to TS 2.9.1 didn't help either